File: EffetAmelioration.cs
namespace exercice_trombone
{
    /// <summary>
    /// Un effet possible d'une amélioration.
    /// </summary>
    internal abstract class EffetAmelioration
    {
        /// <summary>
        /// Valeur modifiée par l'effet
        /// </summary>
        protected int valeur;

        /// <summary>
        /// Applique l'effet sur l'usine.
        /// </summary>
        /// <param name="usine">L'usine qui reçoit l'effet.</param>
        public abstract void AppliquerEffet(Usine usine);
    }
}

File: src/edu/cegepvicto/usine/EffetAmelioration.java
package edu.cegepvicto.usine;

/**
 * Classe qui applique un effet d'amélioration à un usine. La classe
 * doit être spécialisée pour être appliquée à une caractéristique.
 */
public class EffetAmelioration {

    /**
     * Valeur modifiée par l'amélioration.
     */
    private int valeur;

    /**
     * Nom de la caractéristique modifiée.
     */
    private String caracteristique;

    /**
     * Applique un effet à l'usine.
     * @param usine l'usine qui reçoit l'effet d'amélioration.
     */
    public void appliquerEffet(Usine usine) {
        usine.changerCaracteristique(caracteristique, valeur);
    }
}
